packages/monitor-v2/src/llm-bot/common.ts

```diff
@@ -94,63 +193,57 @@ export abstract class OptimisticOracleClient<R extends OptimisticOracleRequest>
-   * @param blockRange The block range to fetch requests from.
-   * @returns A Promise that resolves to an array of OptimisticOracleRequest objects.
+   * Updates the OptimisticOracleClient instance by fetching new Oracle requests updates and storing them in the requests map.
+   * @param blockRange The new blockRange to fetch requests from.
    */
-  protected abstract fetchOracleRequests(blockRange: [number, number]): Promise<R[]>;
+  protected abstract updateOracleRequests(blockRange: BlockRange): Promise<void>;
 
   /**
    * Updates the OptimisticOracleClient instance by fetching new Oracle requests within the specified block range. Returns a new instance.
    * @param blockRange (Optional) The block range to fetch new requests from.
    * @returns A Promise that resolves to a new OptimisticOracleClient instance with updated requests.
    */
-  async updateWithBlockRange(blockRange?: [number, number]): Promise<this> {
-    let range: [number, number];
+  async updateWithBlockRange(blockRange?: BlockRange): Promise<this> {
+    let range: BlockRange;
     if (blockRange) {
-      if (blockRange[0] > blockRange[1]) throw new Error("Invalid block range");
+      if (blockRange[0] > blockRange[1])
+        throw new Error("Start block number should be less than or equal to end block number");
       range = blockRange;
     } else {
       // Calculate the next block range to fetch
       const latestBlock = await this.provider.getBlockNumber();
-      const lastFetchedRange = this.fetchedBlockRanges[this.fetchedBlockRanges.length - 1];
-      const nextStartBlock = lastFetchedRange[1] + 1;
+      const nextStartBlock = this.fetchedBlockRange[1] + 1;
       if (nextStartBlock > latestBlock) return this; // no new blocks to fetch
       range = [nextStartBlock, latestBlock];
     }
     const [startBlock, endBlock] = range;
-    if (
-      this.fetchedBlockRanges.some(([s, e]) => (s <= startBlock && startBlock <= e) || (s <= endBlock && endBlock <= e))
-    )
-      throw new Error("Block range already fetched");
 
-    // Add new range to the list of fetched ranges and sort them by start block number
-    const newRanges = [...this.fetchedBlockRanges, range].sort(([s1], [s2]) => s1 - s2);
+    // Throw an error if the new range doesn't directly follow the last fetched range
+    const lastFetchedEndBlock = this.fetchedBlockRange[this.fetchedBlockRange.length - 1];
+    if (lastFetchedEndBlock != 0 && startBlock !== lastFetchedEndBlock + 1)
+      throw new Error(
+        "New block range does not follow the last fetched block range, there is a gap between the ranges"
+      );
 
-    const newRequests = await this.fetchOracleRequests(range);
+    // We enforce the creation of a new instance of the client to avoid mutating the current instance
+    const newClient = this.copy();
+    newClient.fetchedBlockRange = [newClient.fetchedBlockRange[0], endBlock];
+    await newClient.updateOracleRequests([startBlock, endBlock]);
```

Comment:
I think this is okay as long as the updateOracleRequests is only called here, so the mutation is only during the construction.

However, an alternative way to do this would be to grab the old requests, put them into a new map, update the map, then drop that updated map into the new client. This would allow us to use ReadonlyMap in the client, making it compilation error to modify it after the client is initialized.

Ultimately up to you!

Something like this:
```ts
export abstract class OptimisticOracleClient<R extends OptimisticOracleRequest> {
  protected provider: Provider;

  protected constructor(
    _provider: Provider,
    // Note: you can assign a map to a readonly map, just not vice versa.
    // ReadonlyMap just means you can't modify what's in the map.
    // This is different from making the property readonly, which just protects
    // against assigning a new reference to that variable (replacing the map with
    // a different map).
    readonly requests: ReadonlyMap<string, R> = new Map(),
    readonly fetchedBlockRange: Readonly<BlockRange>
  ) {
    this.provider = _provider;
    this.requests = _requests;
    this.fetchedBlockRange = _fetchedBlockRange;
  }

  ...
  // can be static method call since it wouldn't have to be on an instance.
  // Note: you could do this with a regular method, too. Static just makes it clear that you aren't operating on any instance state other than what's passed in.
  const newRequests = OptimisticOracleClient<R>.updateRequests(provider, this.requests, newBlockRange);

  // Because the new client is pre-loaded with all the data it needs, the map can be ReadonlyMap and the fetchedBlockRange could also be readonly.
  return this.createNewClient(newRequests, [startBlock, endBlock]);
}

Reply:
I like your suggestion, I've added most of the required changes however an abstract function can't be static 🤨 
Are you happy with latest changes or would you like me to dig deeper and make the function static somehow?